File: admin/src/types/northline.ts
/**
 * NorthLine reseller admin DTOs.
 *
 * Mirrors backend.app.schemas.admin_panel.northline. Keep the field names in
 * sync with the Pydantic models — these are JSON-on-the-wire as-is.
 */

export interface NorthlineProfile {
  provider_key: string;
  label: string | null;
  balance_rub: number;
  active: boolean;
  issued_total: number;
  active_total: number;
  spent_rub: number;
}

export interface NorthlinePriceMatrixRow {
  days: number;
  devices: number;
  total_price_rub: number;
  override_key: string | null;
}

export interface NorthlinePrices {
  provider_key: string;
  price_matrix: NorthlinePriceMatrixRow[];
}

export interface NorthlineLtePackage {
  gb: number;
  price_rub: number;
  label: string | null;
  rate: number | null;
}

export interface NorthlineLtePackages {
  packages: NorthlineLtePackage[];
}

export interface NorthlinePriceQuoteParams {
  days: number;
  devices: number;
  unlimited?: boolean;
  lte_gb?: number;
}

export interface NorthlinePriceQuote {
  days: number;
  devices: number;
  tariff_code: string | null;
  total_price_rub: number;
  regular_price_rub: number | null;
  savings_rub: number | null;
  discount_pct: number | null;
  device_discount_pct: number | null;
  term_discount_pct: number | null;
  price_per_device_per_day: number | null;
  traffic_quota_gb: number | null;
  unlimited_traffic: boolean;
  lte_gb: number | null;
  lte_addon_rub: number | null;
  lte_rate_per_gb: number | null;
}

/**
 * NorthLine returns an open-ended dict of branding keys; we type the four
 * we render in the UI but keep the rest reachable for the JSON-debug view.
 */
export interface NorthlineBranding {
  custom_domain?: string | null;
  service_name?: string | null;
  service_description?: string | null;
  support_url?: string | null;
  [key: string]: unknown;
}

export interface NorthlineBrandingResponse {
  branding: NorthlineBranding;
}

export interface NorthlineBrandingUpdateRequest {
  custom_domain?: string | null;
  service_name?: string | null;
  service_description?: string | null;
  support_url?: string | null;
}

File: backend/app/schemas/admin_panel/northline.py
"""Admin → NorthLine reseller schemas.

Thin DTOs over :mod:`app.schemas.northline`. We don't reuse the upstream
classes directly because they carry a few back-compat aliases
(``KeyInfo.devices``, ``Field(validation_alias=...)``) that don't belong
in an admin-facing JSON contract — and because we want the freedom to
add admin-only fields (computed totals, per-row warnings) without
bleeding them into the provider client.
"""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field


class AdminResellerProfile(BaseModel):
    """GET /admin/northline/profile — relayed reseller balance/state."""

    provider_key: str
    label: str | None = None
    balance_rub: float
    active: bool = True
    issued_total: int = 0
    active_total: int = 0
    spent_rub: float = 0.0


class AdminPriceMatrixRow(BaseModel):
    days: int
    devices: int
    total_price_rub: float
    override_key: str | None = None


class AdminResellerPrices(BaseModel):
    """GET /admin/northline/prices."""

    provider_key: str
    price_matrix: list[AdminPriceMatrixRow] = Field(default_factory=list)


class AdminLtePackage(BaseModel):
    gb: int
    price_rub: float
    label: str | None = None
    rate: float | None = None


class AdminLtePackagesResponse(BaseModel):
    """GET /admin/northline/lte-packages."""

    packages: list[AdminLtePackage] = Field(default_factory=list)


class AdminPriceQuote(BaseModel):
    """GET /admin/northline/price-quote — proxied price calculator.

    The bot still uses static prices (per the product decision), but the
    admin UI calls this to estimate one-off custom orders or to verify
    that the upstream price matrix matches our tariff table.
    """

    model_config = ConfigDict(extra="ignore")

    days: int
    devices: int
    tariff_code: str | None = None
    total_price_rub: float
    regular_price_rub: float | None = None
    savings_rub: float | None = None
    discount_pct: int | None = None
    device_discount_pct: int | None = None
    term_discount_pct: int | None = None
    price_per_device_per_day: float | None = None
    traffic_quota_gb: int | None = None
    unlimited_traffic: bool = False
    lte_gb: int | None = None
    lte_addon_rub: float | None = None
    lte_rate_per_gb: float | None = None


class AdminBrandingResponse(BaseModel):
    """GET /admin/northline/branding — current reseller-wide branding."""

    branding: dict[str, Any] = Field(default_factory=dict)


class AdminBrandingUpdateRequest(BaseModel):
    """PUT /admin/northline/branding.

    Every field is optional so the admin can patch one key (say, just the
    support_url) without touching the rest. Sending ``null`` explicitly
    clears the value upstream.
    """

    custom_domain: str | None = Field(default=None, max_length=253)
    service_name: str | None = Field(default=None, max_length=128)
    service_description: str | None = Field(default=None, max_length=512)
    support_url: str | None = Field(default=None, max_length=512)


__all__ = [
    "AdminBrandingResponse",
    "AdminBrandingUpdateRequest",
    "AdminLtePackage",
    "AdminLtePackagesResponse",
    "AdminPriceMatrixRow",
    "AdminPriceQuote",
    "AdminResellerPrices",
    "AdminResellerProfile",
]
